We will be exploring what kind of Neighbourhoods there are in Toronto.

The data in this analysis was taken form:


https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, 

In [1]:
#lets load relevant libraries
!conda install -c anaconda beautifulsoup4 #this works
#!conda install -c conda-forge beautifulsoup4
#install -c conda beautifulsoup4
from bs4 import  BeautifulSoup
import requests
import pandas as pd

Solving environment: done

# All requested packages already installed.



lets find the table

In [2]:
# lets define the url address for extracting information
wiki_url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
#lets download the page
response = requests.get(wiki_url)

In [3]:
# lets parse the dictionary with beautifull soup for extracting the postal codes and their respective names
soup1 = BeautifulSoup(response.text,'html.parser')

In [4]:
table_toronto = soup1.find_all("tbody")[0]

In [5]:
neighBourwardCode = []
boroughList = []
neighBourwardList = []
#new_table = pd.DataFrame(columns=range(162,2), index = [0]) # I know the size 
row_marker = 0
for row in table_toronto.find_all('tr'):
    print("==============")
    #print(row)
    column_marker = 0
    columns = row.find_all('td')
    for column in columns:
        print("..................")
        #print(column.get_text())
        if(column.get_text()[4:7]=='Not'):
            continue
        else:
            neighBourwardCode.append(column.get_text()[1:4])
            text = column.get_text()[4:-2]
            for index,values in enumerate(text):
                if(text[index+1]=='('):
                    boroughList.append(column.get_text()[4:index+5]) # to avoid "\n" characters
                    neighBourwardList.append(column.get_text()[index+6:-2]) # to avoid ")\n" characters
                    if(neighBourwardList[-1][-1]==')'):
                        neighBourwardList[-1] = neighBourwardList[-1][:-1]
                    neighBourwardList[-1] = neighBourwardList[-1].replace('(',' ')
                    neighBourwardList[-1] = neighBourwardList[-1].replace(')','')
                    neighBourwardList[-1] = neighBourwardList[-1].replace(' / ',', ')
                    #for index2, value2 in enumerate(neighBourwardList[-1]):
                    break
                else:
                    continue
            print(neighBourwardCode[-1])
            print(boroughList[-1])
            print(neighBourwardList[-1])
            #new_table.iat[row_marker,column_marker] = column.get_text()
            column_marker += 1
neighBourwardDict = {'Postal Code':neighBourwardCode,'Borough':boroughList,'Neighbourward':neighBourwardList}

..................
..................
..................
M3A
North York
Parkwoods
..................
M4A
North York
Victoria Village
..................
M5A
Downtown Toronto
Regent Park, Harbourfront
..................
M6A
North York
Lawrence Manor, Lawrence Heights
..................
M7A
Queen's Park
Ontario Provincial Government
..................
..................
M9A
Etobicoke
Islington Avenue
..................
M1B
Scarborough
Malvern, Rouge
..................
..................
M3B
North York
Don MillsNorth
..................
M4B
East York
Parkview Hill, Woodbine Gardens
..................
M5B
Downtown Toronto
Garden District, Ryerson
..................
M6B
North York
Glencairn
..................
..................
..................
M9B
Etobicoke
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
..................
M1C
Scarborough
Rouge Hill, Port Union, Highland Creek
..................
..................
M3C
North York
Don MillsSouth Flemingdon Park
.......

Using the dictionary "neighBourwardDict" lets create a data frame

In [6]:
dataF1 = pd.DataFrame(neighBourwardDict)

In [7]:
# lets see what the table looks like
dataF1

Postal Code                                            Borough  \
0           M3A                                         North York   
1           M4A                                         North York   
2           M5A                                   Downtown Toronto   
3           M6A                                         North York   
4           M7A                                       Queen's Park   
..          ...                                                ...   
98          M8X                                          Etobicoke   
99          M4Y                                   Downtown Toronto   
100         M7Y  East TorontoBusiness reply mail Processing Cen...   
101         M8Y                                          Etobicoke   
102         M8Z                                          Etobicoke   

                                         Neighbourward  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Right now we can not see all the rows, so lets specify how many columns and rows we can visualize

In [10]:
pd.set_option('display.max_columns',10)

pd.set_option('display.min_rows',500)

pd.set_option('display.max_rows',1000)

pd.set_option('display.width',200)

pd.set_option('expand_frame_repr',100)

pd.set_option('display.max_colwidth',100)

dataF1

Postal Code                                                       Borough                                                                                        Neighbourward
0           M3A                                                    North York                                                                                            Parkwoods
1           M4A                                                    North York                                                                                     Victoria Village
2           M5A                                              Downtown Toronto                                                                            Regent Park, Harbourfront
3           M6A                                                    North York                                                                     Lawrence Manor, Lawrence Heights
4           M7A                                                  Queen's Park                                                                        Ontario Provincial Government
5           M9A                                                     Etobicoke                                                                                     Islington Avenue
6           M1B                                                   Scarborough                                                                                       Malvern, Rouge
7           M3B                                                    North York                                                                                       Don MillsNorth
8           M4B                                                     East York                                                                      Parkview Hill, Woodbine Gardens
9           M5B                                              Downtown Toronto                                                                             Garden District, Ryerson
10          M6B                                                    North York                                                                                            Glencairn
11          M9B                                                     Etobicoke                               West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
12          M1C                                                   Scarborough                                                               Rouge Hill, Port Union, Highland Creek
13          M3C                                                    North York                                                                       Don MillsSouth Flemingdon Park
14          M4C                                                     East York                                                                                     Woodbine Heights
15          M5C                                              Downtown Toronto                                                                                       St. James Town
16          M6C                                                          York                                                                                   Humewood-Cedarvale
17          M9C                                                     Etobicoke                                        Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
18          M1E                                                   Scarborough                                                                    Guildwood, Morningside, West Hill
19          M4E                                                  East Toronto                                                                                          The Beaches
20          M5E                                              Downtown Toronto                                                                                          Berczy Park
21          M6E                                                   

The data in rows 76, 92and 100 is not clean. Thus lets drop this rows.

In [11]:
dataF1.drop([dataF1.index[76],dataF1.index[92], dataF1.index[100]], inplace=True)

reseting index

In [12]:
dataF1.reset_index(drop=True, inplace=True)

letscheck how many rows we have

In [13]:
dataF1.shape[0]

100

Just in case, lets filter out the rows that have the same postalcode

In [14]:
filtered_data = dataF1.copy(deep=True)

In [15]:
# lets get the postal codes

print('Number of postal codes: ',len(filtered_data['Postal Code'].unique()))

Number of postal codes:  100


So there not repeated postal codes.Then lets print them out

In [16]:
print(filtered_data['Postal Code'].unique())

['M3A' 'M4A' 'M5A' 'M6A' 'M7A' 'M9A' 'M1B' 'M3B' 'M4B' 'M5B' 'M6B' 'M9B'
 'M1C' 'M3C' 'M4C' 'M5C' 'M6C' 'M9C' 'M1E' 'M4E' 'M5E' 'M6E' 'M1G' 'M4G'
 'M5G' 'M6G' 'M1H' 'M2H' 'M3H' 'M4H' 'M5H' 'M6H' 'M1J' 'M2J' 'M3J' 'M4J'
 'M5J' 'M6J' 'M1K' 'M2K' 'M3K' 'M4K' 'M5K' 'M6K' 'M1L' 'M2L' 'M3L' 'M4L'
 'M5L' 'M6L' 'M9L' 'M1M' 'M2M' 'M3M' 'M4M' 'M5M' 'M6M' 'M9M' 'M1N' 'M2N'
 'M3N' 'M4N' 'M5N' 'M6N' 'M9N' 'M1P' 'M2P' 'M4P' 'M5P' 'M6P' 'M9P' 'M1R'
 'M2R' 'M4R' 'M5R' 'M6R' 'M9R' 'M1S' 'M4S' 'M5S' 'M6S' 'M1T' 'M4T' 'M5T'
 'M1V' 'M4V' 'M5V' 'M8V' 'M9V' 'M1W' 'M4W' 'M8W' 'M9W' 'M1X' 'M4X' 'M5X'
 'M8X' 'M4Y' 'M8Y' 'M8Z']


Since there are no repeated postal codes, we do not need to drop any row for this reason

Lets print the new table

In [17]:
filtered_data 

Postal Code                Borough                                                                                        Neighbourward
0          M3A             North York                                                                                            Parkwoods
1          M4A             North York                                                                                     Victoria Village
2          M5A       Downtown Toronto                                                                            Regent Park, Harbourfront
3          M6A             North York                                                                     Lawrence Manor, Lawrence Heights
4          M7A           Queen's Park                                                                        Ontario Provincial Government
5          M9A              Etobicoke                                                                                     Islington Avenue
6          M1B            Scarborough                                                                                       Malvern, Rouge
7          M3B             North York                                                                                       Don MillsNorth
8          M4B              East York                                                                      Parkview Hill, Woodbine Gardens
9          M5B       Downtown Toronto                                                                             Garden District, Ryerson
10         M6B             North York                                                                                            Glencairn
11         M9B              Etobicoke                               West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
12         M1C            Scarborough                                                               Rouge Hill, Port Union, Highland Creek
13         M3C             North York                                                                       Don MillsSouth Flemingdon Park
14         M4C              East York                                                                                     Woodbine Heights
15         M5C       Downtown Toronto                                                                                       St. James Town
16         M6C                   York                                                                                   Humewood-Cedarvale
17         M9C              Etobicoke                                        Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
18         M1E            Scarborough                                                                    Guildwood, Morningside, West Hill
19         M4E           East Toronto                                                                                          The Beaches
20         M5E       Downtown Toronto                                                                                          Berczy Park
21         M6E                   York                                                                                  Caledonia-Fairbanks
22         M1G            Scarborough                                                                                               Woburn
23         M4G              East York                                                                                              Leaside
24         M5G       Downtown Toronto                                                                                   Central Bay Street
25         M6G       Downtown Toronto                                                                                             Christie
26         M1H            Scarborough                                                                                            Cedarbrae
27         M2H             North York                                                                          

Then lets extract the geografic coordinate of the neighbourhoods in the next part

lets load the necessary libraries

In [18]:
%pip install geocoder
import geocoder # import geocoder

Note: you may need to restart the kernel to use updated packages.


In [19]:
#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

In [26]:
geolocator = Nominatim(user_agent='lopezm@gmail.com')

def getCoordinates(df):
    # "df" is the old data frame without coordinates
    #lets create a new dataFrame
    newDataFrame = pd.DataFrame(df,columns = ['Postal Code','Borough','Neighbourward','Latitude','Longitude'])
    #ii = 0
    for idx, place in enumerate(filtered_data['Neighbourward']):
        city = place #'Parkwoods'#'Toronto'
        country = 'Canada'
        location = None
        location = geolocator.geocode(city+','+country)
        if (location != None):
            newDataFrame['Latitude'].loc[idx] = location.latitude
            newDataFrame['Longitude'].loc[idx] = location.longitude            
            print(idx,'The coordinates of {} are {}, {}.'.format(place,newDataFrame['Latitude'].loc[idx], newDataFrame['Longitude'].loc[idx]))
            #
        #
    #
    return newDataFrame
        

dataSet_neighbourwards = getCoordinates(filtered_data)

/opt/conda/envs/Python-3.8-main/lib/python3.8/site-packages/pandas/core/indexing.py:1637: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


0 The coordinates of Parkwoods are 43.7587999, -79.3201966.
1 The coordinates of Victoria Village are 43.732658, -79.3111892.
2 The coordinates of Regent Park, Harbourfront are 43.6400801, -79.3801495.
5 The coordinates of Islington Avenue are 43.6886088, -79.5427878.
6 The coordinates of Malvern, Rouge are 43.8091955, -79.2217008.
10 The coordinates of Glencairn are 43.7087117, -79.4406853.
14 The coordinates of Woodbine Heights are 43.69992, -79.319279.
15 The coordinates of St. James Town are 43.6694032, -79.3727041.
16 The coordinates of Humewood-Cedarvale are 43.6883215, -79.4280805.
19 The coordinates of The Beaches are 43.6710244, -79.296712.
20 The coordinates of Berczy Park are 43.64798435, -79.37539591138858.
21 The coordinates of Caledonia-Fairbanks are 44.415356, -72.020828.
22 The coordinates of Woburn are 43.7598243, -79.2252908.
23 The coordinates of Leaside are 43.7047983, -79.3680904.
24 The coordinates of Central Bay Street are 43.643429850000004, -79.37905948373245.




looking at the index we se that the loop has not taken into account some of them. Lets print the dataframe to check what happens in these rows

lets drop whole row with NaN in "Latitude" column

In [29]:
dataSet_neighbourwards.dropna(subset=["Latitude"], axis=0, inplace=True)
# reset index, because we droped two rows
dataSet_neighbourwards.reset_index(drop=True, inplace=True)
print(dataSet_neighbourwards)

   Postal Code                Borough                   Neighbourward   Latitude   Longitude
0          M3A             North York                       Parkwoods  43.758800  -79.320197
1          M4A             North York                Victoria Village  43.732658  -79.311189
2          M5A       Downtown Toronto       Regent Park, Harbourfront  43.640080  -79.380150
3          M9A              Etobicoke                Islington Avenue  43.688609  -79.542788
4          M1B            Scarborough                  Malvern, Rouge  43.809196  -79.221701
5          M6B             North York                       Glencairn  43.708712  -79.440685
6          M4C              East York                Woodbine Heights  43.699920  -79.319279
7          M5C       Downtown Toronto                  St. James Town  43.669403  -79.372704
8          M6C                   York              Humewood-Cedarvale  43.688322  -79.428080
9          M4E           East Toronto                     The Beaches 

In [30]:
#It looks like geopy got the wrong coordinates for rows 11,16,19,25,36,47 
# Lets drop these lines:
dataSet_neighbourwards.drop([dataSet_neighbourwards.index[16],dataSet_neighbourwards.index[16],dataSet_neighbourwards.index[19],dataSet_neighbourwards.index[25], dataSet_neighbourwards.index[36],dataSet_neighbourwards.index[47]], inplace=True)
# reset index, because we droped two rows
dataSet_neighbourwards.reset_index(drop=True, inplace=True)

After droping the rows with wrong coordinates, lets check the new table:

In [32]:
print("=========================================================")
print("=========================================================")
print("=========================================================")
print("Clean Data Set")
dataSet_neighbourwards

Clean Data Set


Postal Code                Borough                   Neighbourward   Latitude  Longitude
0          M3A             North York                       Parkwoods  43.758800 -79.320197
1          M4A             North York                Victoria Village  43.732658 -79.311189
2          M5A       Downtown Toronto       Regent Park, Harbourfront  43.640080 -79.380150
3          M9A              Etobicoke                Islington Avenue  43.688609 -79.542788
4          M1B            Scarborough                  Malvern, Rouge  43.809196 -79.221701
5          M6B             North York                       Glencairn  43.708712 -79.440685
6          M4C              East York                Woodbine Heights  43.699920 -79.319279
7          M5C       Downtown Toronto                  St. James Town  43.669403 -79.372704
8          M6C                   York              Humewood-Cedarvale  43.688322 -79.428080
9          M4E           East Toronto                     The Beaches  43.671024 -79.296712
10         M5E       Downtown Toronto                     Berczy Park  43.647984 -79.375396
11         M6E                   York             Caledonia-Fairbanks  44.415356 -72.020828
12         M1G            Scarborough                          Woburn  43.759824 -79.225291
13         M4G              East York                         Leaside  43.704798 -79.368090
14         M5G       Downtown Toronto              Central Bay Street  43.643430 -79.379059
15         M6G       Downtown Toronto                        Christie  43.664111 -79.418405
16         M2H             North York               Hillcrest Village  43.681695 -79.425712
17         M4H              East York                Thorncliffe Park  43.704553 -79.345407
18         M6H           West Toronto    Dufferin, Dovercourt Village  43.660202 -79.435719
19         M1J            Scarborough             Scarborough Village  43.743742 -79.211632
20         M4J  East YorkEast Toronto              The Danforth  East  43.688171 -79.301816
21         M2K             North York                 Bayview Village  43.769197 -79.376662
22         M4K           East Toronto    The Danforth West, Riverdale  43.674709 -79.366702
23         M9L             North York                   Humber Summit  43.760078 -79.571760
24         M2M             North York         Willowdale, Newtonbrook  43.789576 -79.417588
25         M4M           East Toronto                 Studio District  43.649210 -79.388026
26         M4N        Central Toronto                   Lawrence Park  43.729199 -79.403252
27         M5N        Central Toronto                        Roselawn  43.706507 -79.421602
28         M6N                   York   Runnymede, The Junction North  43.661070 -79.480327
29         M9N                   York                          Weston  43.700161 -79.516247
30         M2P             North York                 York Mills West  43.744039 -79.406657
31         M4P        Central Toronto                Davisville North  43.697936 -79.397291
32         M5P        Central Toronto        Forest Hill North & West  43.693559 -79.413902
33         M9P              Etobicoke                       Westmount  45.485733 -73.596395
34         M4R        Central Toronto              North Toronto West  43.653482 -79.383935
35         M6R           West Toronto          Parkdale, Roncesvalles  43.641281 -79.451599
36         M1S            Scarborough                      Agincourt   43.785353 -79.278549
37         M4S        Central Toronto                      Davisville  43.697936 -79.397291
38         M5S       Downtown Toronto  University of Toronto, Harbord  43.664096 -79.398669
39         M6S           West Toronto              Runnymede, Swansea  43.651778 -79.475923
40         M4T        Central Toronto     Moore Park, Summerhill East  43.686053 -79.376519
41         M4W       Downtown Toronto                        Rosedale  43.678356 -79.380746
42         M8W              Etobicoke          

In [33]:
#lets make a backup
dataSet_neighbourwards_1 = dataSet_neighbourwards.copy(deep=True)
dataSet_neighbourwards_1

Postal Code                Borough                   Neighbourward   Latitude  Longitude
0          M3A             North York                       Parkwoods  43.758800 -79.320197
1          M4A             North York                Victoria Village  43.732658 -79.311189
2          M5A       Downtown Toronto       Regent Park, Harbourfront  43.640080 -79.380150
3          M9A              Etobicoke                Islington Avenue  43.688609 -79.542788
4          M1B            Scarborough                  Malvern, Rouge  43.809196 -79.221701
5          M6B             North York                       Glencairn  43.708712 -79.440685
6          M4C              East York                Woodbine Heights  43.699920 -79.319279
7          M5C       Downtown Toronto                  St. James Town  43.669403 -79.372704
8          M6C                   York              Humewood-Cedarvale  43.688322 -79.428080
9          M4E           East Toronto                     The Beaches  43.671024 -79.296712
10         M5E       Downtown Toronto                     Berczy Park  43.647984 -79.375396
11         M6E                   York             Caledonia-Fairbanks  44.415356 -72.020828
12         M1G            Scarborough                          Woburn  43.759824 -79.225291
13         M4G              East York                         Leaside  43.704798 -79.368090
14         M5G       Downtown Toronto              Central Bay Street  43.643430 -79.379059
15         M6G       Downtown Toronto                        Christie  43.664111 -79.418405
16         M2H             North York               Hillcrest Village  43.681695 -79.425712
17         M4H              East York                Thorncliffe Park  43.704553 -79.345407
18         M6H           West Toronto    Dufferin, Dovercourt Village  43.660202 -79.435719
19         M1J            Scarborough             Scarborough Village  43.743742 -79.211632
20         M4J  East YorkEast Toronto              The Danforth  East  43.688171 -79.301816
21         M2K             North York                 Bayview Village  43.769197 -79.376662
22         M4K           East Toronto    The Danforth West, Riverdale  43.674709 -79.366702
23         M9L             North York                   Humber Summit  43.760078 -79.571760
24         M2M             North York         Willowdale, Newtonbrook  43.789576 -79.417588
25         M4M           East Toronto                 Studio District  43.649210 -79.388026
26         M4N        Central Toronto                   Lawrence Park  43.729199 -79.403252
27         M5N        Central Toronto                        Roselawn  43.706507 -79.421602
28         M6N                   York   Runnymede, The Junction North  43.661070 -79.480327
29         M9N                   York                          Weston  43.700161 -79.516247
30         M2P             North York                 York Mills West  43.744039 -79.406657
31         M4P        Central Toronto                Davisville North  43.697936 -79.397291
32         M5P        Central Toronto        Forest Hill North & West  43.693559 -79.413902
33         M9P              Etobicoke                       Westmount  45.485733 -73.596395
34         M4R        Central Toronto              North Toronto West  43.653482 -79.383935
35         M6R           West Toronto          Parkdale, Roncesvalles  43.641281 -79.451599
36         M1S            Scarborough                      Agincourt   43.785353 -79.278549
37         M4S        Central Toronto                      Davisville  43.697936 -79.397291
38         M5S       Downtown Toronto  University of Toronto, Harbord  43.664096 -79.398669
39         M6S           West Toronto              Runnymede, Swansea  43.651778 -79.475923
40         M4T        Central Toronto     Moore Park, Summerhill East  43.686053 -79.376519
41         M4W       Downtown Toronto                        Rosedale  43.678356 -79.380746
42         M8W              Etobicoke          

Lets explore the neighbourwards in Toronto. 

In [35]:
import json # this package handles JSON files
from pandas.io.json import json_normalize # for tranforming JSON files into a pandas dataframes
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans # k-means from clustering

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

import math as m # for using the isnan method

Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python-3.8-main

  added / updated specs:
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    altair-4.1.0               |             py_1         614 KB  conda-forge
    branca-0.4.2               |     pyhd8ed1ab_0          26 KB  conda-forge
    certifi-2021.5.30          |   py38h578d9bd_0         141 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    openssl-1.1.1k             |       h7f98852_0         2.1 MB  conda-forge
    python_abi-3.8             |           2_cp38           4 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    widgetsnbextension-3.5.1   |   py38h578d9bd_4         1.8 MB  conda-for

After loading the libraries, lets continue. Next, we are going to start utilizing the Foursquare API to explore the Neighbourwards and segment them.

In [36]:
CLIENT_SECRET = '3VHKVC3MXM4AAFG0IT5COU0JUN5T2KRTS25AURQXQ0WJA5WL' # your Foursquare Secret
CLIENT_ID = 'I02STQSDKHLNOKIQQWU0DIIRTKDJLTQ2ISDVKMLVAKO1ZR1V' # your Foursquare ID

VERSION = '20180605' # Foursquare API version
ACCESS_TOKEN = 'O3UDS4UFECIZSYP4LSDAOX2KOSDMOWI0LKTUR4MS4SFPP5IZ' # your FourSquare Access Token


LIMIT = dataSet_neighbourwards.shape[0] #100 # A default Foursquare API limit value

Now, let's get the top 100 venues within a  certain radius

Lets define the get request URL that forsquare is going to use. As an examaple, lets see what we can findin the first neighbouhood of our list

In [38]:
ii = 0
neighborward_latitude = dataSet_neighbourwards['Latitude'].loc[ii]
neighborward_longitude = dataSet_neighbourwards['Longitude'].loc[ii]

radius1 = 500 # define radius
LIMIT = dataSet_neighbourwards.shape[0] #100 # A default Foursquare API limit value
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, neighborward_latitude,neighborward_longitude, VERSION, radius1, LIMIT)
print(url)

https://api.foursquare.com/v2/venues/explore?client_id=I02STQSDKHLNOKIQQWU0DIIRTKDJLTQ2ISDVKMLVAKO1ZR1V&client_secret=3VHKVC3MXM4AAFG0IT5COU0JUN5T2KRTS25AURQXQ0WJA5WL&ll=43.7587999,-79.3201966&v=20180605&radius=500&limit=45


Lets use the GET request and examine the information that comes back

In [39]:
results = requests.get(url).json()
print(results)

{'meta': {'code': 200, 'requestId': '61215bb0ca77135544933d3d'}, 'response': {'suggestedFilters': {'header': 'Tap to show:', 'filters': [{'name': 'Open now', 'key': 'openNow'}]}, 'headerLocation': 'Parkwoods - Donalda', 'headerFullLocation': 'Parkwoods - Donalda, Toronto', 'headerLocationGranularity': 'neighborhood', 'totalResults': 13, 'suggestedBounds': {'ne': {'lat': 43.7632999045, 'lng': -79.31397776399336}, 'sw': {'lat': 43.7542998955, 'lng': -79.32641543600664}}, 'groups': [{'type': 'Recommended Places', 'name': 'recommended', 'items': [{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4b8991cbf964a520814232e3', 'name': "Allwyn's Bakery", 'location': {'address': '81 Underhill drive', 'lat': 43.75984035203157, 'lng': -79.32471879917513, 'labeledLatLngs': [{'label': 'display', 'lat': 43.75984035203157, 'lng': -79.32471879917513}], 'distance': 381, 'postalCode': 'M3A 1Z5', 'cc': 'CA'

In [40]:
# this function extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

lets clean the jason file

In [42]:
venues = results['response']['groups'][0]['items']
    
venuesList = pd.json_normalize(venues) # flatten JSON

#lets make a backup copy
venuesListCopy1 = venuesList.copy(deep=True)

filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
venuesList = venuesList.loc[:, filtered_columns]

venuesListCopy1 = venuesList.copy(deep=True)

# filter the category for each row
venuesList['venue.categories'] = venuesList.apply(get_category_type, axis=1)

# clean columns
venuesList.columns = [col.split(".")[-1] for col in venuesList.columns]

venuesList

name            categories        lat        lng
0            Allwyn's Bakery  Caribbean Restaurant  43.759840 -79.324719
1                       LCBO          Liquor Store  43.757774 -79.314257
2         Shoppers Drug Mart              Pharmacy  43.760857 -79.324961
3               Petro-Canada           Gas Station  43.757950 -79.315187
4            TD Canada Trust                  Bank  43.757440 -79.314838
5                Pizza Pizza           Pizza Place  43.760231 -79.325666
6     Parkwoods Coin Laundry       Laundry Service  43.760386 -79.324894
7           Family Food Fair                   ATM  43.760422 -79.325012
8   Parkwoods Village Centre         Shopping Mall  43.760735 -79.324873
9                  Dollarama        Discount Store  43.760341 -79.325519
10                    Rogers     Electronics Store  43.758737 -79.314383
11                  La Notre           Coffee Shop  43.760704 -79.325396
12       Spicy Chicken House    Chinese Restaurant  43.760639 -79.325671

So this are the stores in the first neighbourhood

And how many venues were returned by Foursquare?

In [44]:
print('{} venues were returned by Foursquare.'.format(venuesList.shape[0]))

13 venues were returned by Foursquare.


Now lets repeat this process for all the neighbouhoods of our list. For this porpouse, I will define the following function:

In [45]:
def getVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

### CHECK
    venuesList = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    venuesList.columns = ['Neighbourward', 
                  'Neighbourward Latitude', 
                  'Neighbourward Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(venuesList)

Now lets get all the venues from all the neighbouhoods

In [46]:
cityVenues = getVenues(names=dataSet_neighbourwards['Neighbourward'],
                                   latitudes=dataSet_neighbourwards['Latitude'],
                                   longitudes=dataSet_neighbourwards['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Islington Avenue
Malvern, Rouge
Glencairn
Woodbine Heights
St. James Town
Humewood-Cedarvale
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Hillcrest Village
Thorncliffe Park
Dufferin, Dovercourt Village
Scarborough Village
The Danforth  East
Bayview Village
The Danforth West, Riverdale
Humber Summit
Willowdale, Newtonbrook
Studio District
Lawrence Park
Roselawn
Runnymede, The Junction North
Weston
York Mills West
Davisville North
Forest Hill North & West
Westmount
North Toronto West
Parkdale, Roncesvalles
Agincourt 
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Rosedale
Alderwood, Long Branch
St. James Town, Cabbagetown
Church and Wellesley


Just in case, I am making a copy

In [47]:
cityVenues_bc = cityVenues.copy(deep=True)

These are the first 40 venues in our list

In [48]:
cityVenues.head(40)

Neighbourward  Neighbourward Latitude  Neighbourward Longitude                          Venue  Venue Latitude  Venue Longitude             Venue Category
0                   Parkwoods               43.758800               -79.320197                Allwyn's Bakery       43.759840       -79.324719       Caribbean Restaurant
1                   Parkwoods               43.758800               -79.320197                           LCBO       43.757774       -79.314257               Liquor Store
2                   Parkwoods               43.758800               -79.320197             Shoppers Drug Mart       43.760857       -79.324961                   Pharmacy
3                   Parkwoods               43.758800               -79.320197                   Petro-Canada       43.757950       -79.315187                Gas Station
4                   Parkwoods               43.758800               -79.320197                TD Canada Trust       43.757440       -79.314838                       Bank
5                   Parkwoods               43.758800               -79.320197                    Pizza Pizza       43.760231       -79.325666                Pizza Place
6                   Parkwoods               43.758800               -79.320197         Parkwoods Coin Laundry       43.760386       -79.324894            Laundry Service
7                   Parkwoods               43.758800               -79.320197               Family Food Fair       43.760422       -79.325012                        ATM
8                   Parkwoods               43.758800               -79.320197       Parkwoods Village Centre       43.760735       -79.324873              Shopping Mall
9                   Parkwoods               43.758800               -79.320197                      Dollarama       43.760341       -79.325519             Discount Store
10                  Parkwoods               43.758800               -79.320197                         Rogers       43.758737       -79.314383          Electronics Store
11                  Parkwoods               43.758800               -79.320197                       La Notre       43.760704       -79.325396                Coffee Shop
12                  Parkwoods               43.758800               -79.320197            Spicy Chicken House       43.760639       -79.325671         Chinese Restaurant
13           Victoria Village               43.732658               -79.311189                        Jatujak       43.736208       -79.307668            Thai Restaurant
14           Victoria Village               43.732658               -79.311189               Armenian Kitchen       43.731071       -79.305390  Middle Eastern Restaurant
15           Victoria Village               43.732658               -79.311189                   Wigmore Park       43.731023       -79.310771                       Park
16           Victoria Village               43.732658               -79.311189  TTC Bus #24 Victoria Park Ave       43.735197       -79.306977                   Bus Line
17           Victoria Village               43.732658               -79.311189                        Tagpuan       43.735943       -79.307651           Asian Restaurant
18  Regent Park, Harbourfront               43.640080               -79.380150            Harbour Square Park       43.639253       -79.378395                       Park
19  Regent Park, Harbourfront               43.640080               -79.380150                   Harbourfront       43.639526       -79.380688               Neighborhood
20  Regent Park, Harbourfront               43.640080               -79.380150                   Lake Ontario       43.638945       -79.379665                       Lake
21  Regent Park, Harbourfront               43.640080               -79.380150                    BeaverTails       43.639736       -79.380068               Dessert Shop
22  Regent Park, Harbourfront               43.640080               -79.380150                           M

lets study how many venues were found in each district

In [49]:
cityVenues.groupby('Neighbourward').count()

Neighbourward Latitude  Neighbourward Longitude  Venue  Venue Latitude  Venue Longitude  Venue Category
Neighbourward                                                                                                                          
Agincourt                                           13                       13     13              13               13              13
Alderwood, Long Branch                               6                        6      6               6                6               6
Bayview Village                                     12                       12     12              12               12              12
Berczy Park                                         45                       45     45              45               45              45
Caledonia-Fairbanks                                  7                        7      7               7                7               7
Central Bay Street                                  45                       45     45              45               45              45
Christie                                            45                       45     45              45               45              45
Church and Wellesley                                45                       45     45              45               45              45
Davisville                                          40                       40     40              40               40              40
Davisville North                                    40                       40     40              40               40              40
Dufferin, Dovercourt Village                        45                       45     45              45               45              45
Forest Hill North & West                             5                        5      5               5                5               5
Glencairn                                            6                        6      6               6                6               6
Hillcrest Village                                   43                       43     43              43               43              43
Humber Summit                                        4                        4      4               4                4               4
Humewood-Cedarvale                                   6                        6      6               6                6               6
Islington Avenue                                     3                        3      3               3                3               3
Lawrence Park                                       45                       45     45              45               45              45
Leaside                                              4                        4      4               4                4               4
Malvern, Rouge                                      13                       13     13              13               13              13
Moore Park, Summerhill East                          5                        5      5               5                5               5
North Toronto West                                  45                       45     45              45               45              45
Parkdale, Roncesvalles                              27                       27     27              27               27              27
Parkwoods                                           13                       13     13              13               13              13
Regent Park, Harbourfront                           45                       45     45              45               45              45
Rosedale                                             5                        5      5               5                5               5
Roselawn                                             2                        2      2               2                2               2
Runnymede, Swansea                                  42                       42     42  

In [50]:
print('There are {} uniques categories.'.format(len(cityVenues['Venue Category'].unique()))) # There are 272 uniques

There are 222 uniques categories.


No w I am analyzing what are the most popular store in each district

In [51]:
hotspotFrame = pd.get_dummies(cityVenues[['Venue Category']], prefix="", prefix_sep="")

In [52]:
hotspotFrameCopy1 = hotspotFrame.copy(deep=True)

In [53]:
# return neighborward column back to our dataframe
hotspotFrame['Neighbourward'] = cityVenues['Neighbourward'] 

In [55]:
# for safety lets save a copy
hotspotFrameCopy2 = hotspotFrame.copy(deep=True)


In [56]:
# swapping neighborward column with the first column
columnsTemp = [hotspotFrame.columns[-1]] + list(hotspotFrame.columns[:-1])
hotspotFrame = hotspotFrame[columnsTemp]

Now lets check the modified data Frame, s in the lab example

In [58]:
hotspotFrame

Neighbourward  ATM  Afghan Restaurant  African Restaurant  American Restaurant  ...  Vietnamese Restaurant  Warehouse Store  Wine Bar  Women's Store  Yoga Studio
0                          Parkwoods    0                  0                   0                    0  ...                      0                0         0              0            0
1                          Parkwoods    0                  0                   0                    0  ...                      0                0         0              0            0
2                          Parkwoods    0                  0                   0                    0  ...                      0                0         0              0            0
3                          Parkwoods    0                  0                   0                    0  ...                      0                0         0              0            0
4                          Parkwoods    0                  0                   0                    0  ...                      0                0         0              0            0
5                          Parkwoods    0                  0                   0                    0  ...                      0                0         0              0            0
6                          Parkwoods    0                  0                   0                    0  ...                      0                0         0              0            0
7                          Parkwoods    1                  0                   0                    0  ...                      0                0         0              0            0
8                          Parkwoods    0                  0                   0                    0  ...                      0                0         0              0            0
9                          Parkwoods    0                  0                   0                    0  ...                      0                0         0              0            0
10                         Parkwoods    0                  0                   0                    0  ...                      0                0         0              0            0
11                         Parkwoods    0                  0                   0                    0  ...                      0                0         0              0            0
12                         Parkwoods    0                  0                   0                    0  ...                      0                0         0              0            0
13                  Victoria Village    0                  0                   0                    0  ...                      0                0         0              0            0
14                  Victoria Village    0                  0                   0                    0  ...                      0                0         0              0            0
15                  Victoria Village    0                  0                   0                    0  ...                      0                0         0              0            0
16                  Victoria Village    0                  0                   0                    0  ...                      0                0         0              0            0
17                  Victoria Village    0                  0                   0                    0  ...                      0                0         0              0            0
18         Regent Park, Harbourfront    0                  0                   0                    0  ...                      0                0         0              0            0
19         Regent Park, Harbourfront    0                  0                   0                    0  ...                      0                0         0              0            0
20         Regent Park, Harbourfront    0                  0                   0                    0  ...                      0         

Now lets chek how many ros and columns this dataframe has

In [59]:
print(hotspotFrame.shape) # (2110, 273)

(1048, 223)


So there are 2110 venues (rows) and 223 categories of venues (columns)

Now lets calculate how may venues of each type there are on average in each Neighbouhood

In [60]:
groupedVenues = hotspotFrame.groupby('Neighbourward').mean().reset_index()

In [62]:
groupedVenues

Neighbourward       ATM  Afghan Restaurant  African Restaurant  American Restaurant  ...  Vietnamese Restaurant  Warehouse Store  Wine Bar  Women's Store  Yoga Studio
0                       Agincourt   0.000000           0.000000            0.000000             0.000000  ...               0.076923         0.000000  0.000000       0.000000     0.000000
1           Alderwood, Long Branch  0.000000           0.000000            0.000000             0.000000  ...               0.000000         0.000000  0.000000       0.000000     0.000000
2                  Bayview Village  0.000000           0.000000            0.000000             0.000000  ...               0.000000         0.000000  0.000000       0.000000     0.000000
3                      Berczy Park  0.000000           0.000000            0.000000             0.000000  ...               0.000000         0.000000  0.000000       0.000000     0.022222
4              Caledonia-Fairbanks  0.000000           0.000000            0.000000             0.000000  ...               0.000000         0.000000  0.000000       0.000000     0.000000
5               Central Bay Street  0.000000           0.000000            0.000000             0.000000  ...               0.000000         0.000000  0.000000       0.000000     0.000000
6                         Christie  0.000000           0.000000            0.000000             0.022222  ...               0.022222         0.000000  0.022222       0.000000     0.000000
7             Church and Wellesley  0.000000           0.000000            0.000000             0.000000  ...               0.000000         0.000000  0.000000       0.000000     0.044444
8                       Davisville  0.000000           0.000000            0.000000             0.000000  ...               0.000000         0.000000  0.000000       0.000000     0.000000
9                 Davisville North  0.000000           0.000000            0.000000             0.000000  ...               0.000000         0.000000  0.000000       0.000000     0.000000
10    Dufferin, Dovercourt Village  0.000000           0.000000            0.022222             0.000000  ...               0.022222         0.000000  0.000000       0.000000     0.000000
11        Forest Hill North & West  0.000000           0.000000            0.000000             0.000000  ...               0.000000         0.000000  0.000000       0.000000     0.000000
12                       Glencairn  0.000000           0.000000            0.000000             0.000000  ...               0.000000         0.000000  0.000000       0.000000     0.000000
13               Hillcrest Village  0.000000           0.000000            0.000000             0.023256  ...               0.000000         0.000000  0.000000       0.000000     0.000000
14                   Humber Summit  0.000000           0.000000            0.000000             0.000000  ...               0.000000         0.000000  0.000000       0.000000     0.000000
15              Humewood-Cedarvale  0.000000           0.000000            0.000000             0.000000  ...               0.000000         0.000000  0.000000       0.000000     0.000000
16                Islington Avenue  0.000000           0.000000            0.000000             0.000000  ...               0.000000         0.000000  0.000000       0.000000     0.000000
17                   Lawrence Park  0.000000           0.000000            0.000000             0.022222  ...               0.000000         0.000000  0.000000       0.000000     0.000000
18                         Leaside  0.000000           0.000000            0.000000             0.000000  ...               0.000000         0.000000  0.000000       0.000000     0.000000
19                  Malvern, Rouge  0.000000           0.000000            0.000000             0.000000  ...               0.000000         0.000000  0.000000       0.000000     0.000000
20     Moore Park, Summerhill East  0.000000           0.000000          

Lets check the size of the data frame 

In [64]:
print(groupedVenues.shape)

(45, 223)


This means that we have found 223 types of venues in 45 different neighbouhoods

Next, I am going to list each neighbourhood along with the top 5 most common venues

I will follow a similar procedure as in the lab example

In [65]:
number_mostFrequeest_venues = 5
for ward in groupedVenues['Neighbourward']:
    print("===-"+ward+"===")
    temp = groupedVenues[groupedVenues['Neighbourward'] == ward].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float) # converting values to float
    temp = temp.round({'freq': 2}) # rounding to 2 decimal digits
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(number_mostFrequeest_venues)) #printing values with the fre  uency in ascing order
    print('\n')

===-Agincourt ===
                 venue  freq
0   Chinese Restaurant  0.23
1     Asian Restaurant  0.15
2  Rental Car Location  0.08
3        Train Station  0.08
4    Korean Restaurant  0.08


===-Alderwood, Long Branch===
            venue  freq
0  Sandwich Place  0.17
1     Pizza Place  0.17
2             Pub  0.17
3     Gas Station  0.17
4           Trail  0.17


===-Bayview Village===
                 venue  freq
0                 Bank  0.17
1       Sandwich Place  0.08
2   Persian Restaurant  0.08
3          Fish Market  0.08
4  Sporting Goods Shop  0.08


===-Berczy Park===
                 venue  freq
0                 Café  0.09
1  Japanese Restaurant  0.07
2                Hotel  0.07
3             Beer Bar  0.04
4         Cocktail Bar  0.04


===-Caledonia-Fairbanks===
                    venue  freq
0           Deli / Bodega  0.14
1                    Bank  0.14
2          Private School  0.14
3          Sandwich Place  0.14
4  Furniture / Home Store  0.14


===-Central Bay

Let's integrate this data into a pandas dataframe. For this purpose, the following function sorts the venues in descending order.

In [66]:
def return_most_common_venues(row, number_mostFrequeest_venues):
    rowCategorieType = row.iloc[1:] # we are dropping the neighbouward name and only are selecting the venue fecuency
    rowCategorieType_sorted = rowCategorieType.sort_values(ascending=False)
    
    return rowCategorieType_sorted.index.values[0:number_mostFrequeest_venues]

Now let's define a new dataframe and display the most common 10 venues for each neighborward.

In [67]:
number_mostFrequeest_venues = 10

indicators = ['st', 'nd', 'rd']

# we need tocreate columns to fit the number of top venues
import numpy as np
columns = ['Neighbourward']
for ind in np.arange(number_mostFrequeest_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))
        
# and now is when I define a new dataframe
wards_venues_sorted = pd.DataFrame(columns=columns)

The next step is to fill the dataframe with data

In [68]:
wards_venues_sorted['Neighbourward'] = groupedVenues['Neighbourward']
for ind in np.arange(groupedVenues.shape[0]):
    wards_venues_sorted.iloc[ind, 1:] = return_most_common_venues(groupedVenues.iloc[ind, :], number_mostFrequeest_venues)

In [69]:
wards_venues_sorted

Neighbourward       1st Most Common Venue      2nd Most Common Venue        3rd Most Common Venue 4th Most Common Venue  ...          6th Most Common Venue  \
0                       Agincourt           Chinese Restaurant           Asian Restaurant          Rental Car Location         Train Station  ...           Cantonese Restaurant   
1           Alderwood, Long Branch              Sandwich Place                Pizza Place                          Pub           Gas Station  ...                            Gym   
2                  Bayview Village                        Bank             Sandwich Place           Persian Restaurant           Fish Market  ...                 Breakfast Spot   
3                      Berczy Park                        Café        Japanese Restaurant                        Hotel              Beer Bar  ...             Seafood Restaurant   
4              Caledonia-Fairbanks               Deli / Bodega                       Bank               Private School        Sandwich Place  ...                  Movie Theater   
5               Central Bay Street                       Hotel                Coffee Shop          Japanese Restaurant                  Park  ...                           Café   
6                         Christie           Korean Restaurant                Coffee Shop                         Café       Bubble Tea Shop  ...                   Dessert Shop   
7             Church and Wellesley                     Gay Bar                Coffee Shop                  Yoga Studio      Sushi Restaurant  ...                     Restaurant   
8                       Davisville          Italian Restaurant           Sushi Restaurant               Ice Cream Shop                   Pub  ...              Convenience Store   
9                 Davisville North          Italian Restaurant           Sushi Restaurant               Ice Cream Shop                   Pub  ...              Convenience Store   
10    Dufferin, Dovercourt Village                         Bar                     Bakery                  Coffee Shop            Beer Store  ...                           Café   
11        Forest Hill North & West                        Park                 Playground                 Home Service                  Bank  ...                    Pizza Place   
12                       Glencairn         Japanese Restaurant                 Playground                  Pizza Place      Asian Restaurant  ...                         Bakery   
13               Hillcrest Village          Italian Restaurant          Indian Restaurant           Mexican Restaurant          Burger Joint  ...                         Bakery   
14                   Humber Summit  Construction & Landscaping                     Bakery                  Flower Shop                  Park  ...  Paper / Office Supplies Store   
15              Humewood-Cedarvale                       Field             Soccer Stadium                      Dog Run     Convenience Store  ...                   Hockey Arena   
16                Islington Avenue             Supplement Shop                   Pharmacy               Baseball Field                   ATM  ...                         Museum   
17                   Lawrence Park            Sushi Restaurant                     Bakery                  Coffee Shop    Italian Restaurant  ...                 Cosmetics Shop   
18                         Leaside                        Park             Sandwich Place          Japanese Restaurant     Convenience Store  ...                     Nail Salon   
19                  Malvern, Rouge        Fast Food Restaurant                Pizza Place                     Pharmacy        Sandwich Place  ...                   Skating Rink   
20     Moore Park, Summerhill East                        Park                 Playground                  Candy Store         Grocery Store  ...           Outdoor Supply Store   
21              North Toronto West              Clothing Stor

 ===============================================

===============================================

===============================================

===============================================

===============================================

Now I can start clustering the  Neighbourhoods

I am going to use the k-means algortihm to cluster the neighbouhoods into 4 categories

In [70]:
kclusters = 4 # number of clusters
groupedVenues_clustering1 = groupedVenues.drop('Neighbourward', 1)
groupedVenues_clustering1.head(20)

ATM  Afghan Restaurant  African Restaurant  American Restaurant  Art Gallery  ...  Vietnamese Restaurant  Warehouse Store  Wine Bar  Women's Store  Yoga Studio
0   0.0                0.0            0.000000             0.000000     0.000000  ...               0.076923              0.0  0.000000            0.0     0.000000
1   0.0                0.0            0.000000             0.000000     0.000000  ...               0.000000              0.0  0.000000            0.0     0.000000
2   0.0                0.0            0.000000             0.000000     0.000000  ...               0.000000              0.0  0.000000            0.0     0.000000
3   0.0                0.0            0.000000             0.000000     0.022222  ...               0.000000              0.0  0.000000            0.0     0.022222
4   0.0                0.0            0.000000             0.000000     0.000000  ...               0.000000              0.0  0.000000            0.0     0.000000
5   0.0                0.0            0.000000             0.000000     0.000000  ...               0.000000              0.0  0.000000            0.0     0.000000
6   0.0                0.0            0.000000             0.022222     0.000000  ...               0.022222              0.0  0.022222            0.0     0.000000
7   0.0                0.0            0.000000             0.000000     0.000000  ...               0.000000              0.0  0.000000            0.0     0.044444
8   0.0                0.0            0.000000             0.000000     0.000000  ...               0.000000              0.0  0.000000            0.0     0.000000
9   0.0                0.0            0.000000             0.000000     0.000000  ...               0.000000              0.0  0.000000            0.0     0.000000
10  0.0                0.0            0.022222             0.000000     0.000000  ...               0.022222              0.0  0.000000            0.0     0.000000
11  0.0                0.0            0.000000             0.000000     0.000000  ...               0.000000              0.0  0.000000            0.0     0.000000
12  0.0                0.0            0.000000             0.000000     0.000000  ...               0.000000              0.0  0.000000            0.0     0.000000
13  0.0                0.0            0.000000             0.023256     0.023256  ...               0.000000              0.0  0.000000            0.0     0.000000
14  0.0                0.0            0.000000             0.000000     0.000000  ...               0.000000              0.0  0.000000            0.0     0.000000
15  0.0                0.0            0.000000             0.000000     0.000000  ...               0.000000              0.0  0.000000            0.0     0.000000
16  0.0                0.0            0.000000             0.000000     0.000000  ...               0.000000              0.0  0.000000            0.0     0.000000
17  0.0                0.0            0.000000             0.022222     0.000000  ...               0.000000              0.0  0.000000            0.0     0.000000
18  0.0                0.0            0.000000             0.000000     0.000000  ...               0.000000              0.0  0.000000            0.0     0.000000
19  0.0                0.0            0.000000             0.000000     0.000000  ...               0.000000              0.0  0.000000            0.0     0.000000

[20 rows x 222 columns]

lets run theclustering algorithm

In [71]:
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(groupedVenues_clustering1)

NowI check the labels generated for each row in the dataframe

In [72]:
print(kmeans.labels_[0:10] )

[0 0 0 0 0 0 0 0 0 0]


Now I generate a new dataframe that list the top 10 venues for each neighbourhood incluing the custer labels

In [73]:
wards_venues_sorted.insert(0, 'Labels for city cluster map', kmeans.labels_)
wards_venues_sortedCopy = wards_venues_sorted.copy(deep=True)
wards_venues_sortedCopy

Labels for city cluster map                   Neighbourward       1st Most Common Venue      2nd Most Common Venue        3rd Most Common Venue  ...          6th Most Common Venue  \
0                             0                      Agincourt           Chinese Restaurant           Asian Restaurant          Rental Car Location  ...           Cantonese Restaurant   
1                             0          Alderwood, Long Branch              Sandwich Place                Pizza Place                          Pub  ...                            Gym   
2                             0                 Bayview Village                        Bank             Sandwich Place           Persian Restaurant  ...                 Breakfast Spot   
3                             0                     Berczy Park                        Café        Japanese Restaurant                        Hotel  ...             Seafood Restaurant   
4                             0             Caledonia-Fairbanks               Deli / Bodega                       Bank               Private School  ...                  Movie Theater   
5                             0              Central Bay Street                       Hotel                Coffee Shop          Japanese Restaurant  ...                           Café   
6                             0                        Christie           Korean Restaurant                Coffee Shop                         Café  ...                   Dessert Shop   
7                             0            Church and Wellesley                     Gay Bar                Coffee Shop                  Yoga Studio  ...                     Restaurant   
8                             0                      Davisville          Italian Restaurant           Sushi Restaurant               Ice Cream Shop  ...              Convenience Store   
9                             0                Davisville North          Italian Restaurant           Sushi Restaurant               Ice Cream Shop  ...              Convenience Store   
10                            0    Dufferin, Dovercourt Village                         Bar                     Bakery                  Coffee Shop  ...                           Café   
11                            1        Forest Hill North & West                        Park                 Playground                 Home Service  ...                    Pizza Place   
12                            0                       Glencairn         Japanese Restaurant                 Playground                  Pizza Place  ...                         Bakery   
13                            0               Hillcrest Village          Italian Restaurant          Indian Restaurant           Mexican Restaurant  ...                         Bakery   
14                            1                   Humber Summit  Construction & Landscaping                     Bakery                  Flower Shop  ...  Paper / Office Supplies Store   
15                            3              Humewood-Cedarvale                       Field             Soccer Stadium                      Dog Run  ...                   Hockey Arena   
16                            0                Islington Avenue             Supplement Shop                   Pharmacy               Baseball Field  ...                         Museum   
17                            0                   Lawrence Park            Sushi Restaurant                     Bakery                  Coffee Shop  ...                 Cosmetics Shop   
18                            1                         Leaside                        Park             Sandwich Place          Japanese Restaurant  ...                     Nail Salon   
19                            0                  Malvern, Rouge        Fast Food Restaurant                Pizza Place                     Pharmacy  ...                   Skating Rink   
20                            1     Moore Park, Summerhill East              

Now lets make a copy

In [74]:
cityVenues_merged = dataSet_neighbourwards.copy(deep=True)

Now I will make onther copy just in case 

In [87]:
cityVenues_merged2 = cityVenues_merged.copy(deep=True)

Now I will join this dataseta above with dataSet_neighbourwards that contains latitude/longitude coordinates for each neighborward

In [88]:
cityVenues_merged2 = cityVenues_merged2.join(wards_venues_sorted.set_index('Neighbourward'), on='Neighbourward')

In [89]:
# checking if there is any nan values
for ff,val in enumerate(cityVenues_merged2['Labels for city cluster map']):

    if(m.isnan(val)):

        print(ff,cityVenues_merged2.loc[ff,'Labels for city cluster map'])

        #cityVenues_merged3.drop([ff],axis=0)

    else:
        pass
        #cityVenues_merged5.loc[:,'Labels for city cluster map'] = cityVenues_merged6.loc[:,'Labels for city cluster map']


Now let's plot the resulting clusters into a map like int he lab exercise

In [90]:
ii = 10
neighborward_latitude = dataSet_neighbourwards['Latitude'].loc[ii]
neighborward_longitude = dataSet_neighbourwards['Longitude'].loc[ii]
latitude  = neighborward_latitude 
longitude = neighborward_longitude
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(cityVenues_merged2['Latitude'], cityVenues_merged2['Longitude'], cityVenues_merged2['Neighbourward'], cityVenues_merged2['Labels for city cluster map']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       


In [92]:
map_clusters

In the map above there are 4 kinds of labels under  the following colours: red, blue, cian and light yelow

===========================================================

===========================================================

===========================================================

===========================================================

===========================================================

===========================================================

Finally I am going to investigate what are the most poplar venues in each cluster

I am interested in knowing what kind of venues set each cluster apart 

In [94]:
print("Now lets examine cluster 1")

cityVenues_merged2.loc[cityVenues_merged2['Labels for city cluster map'] == 0, cityVenues_merged2.columns[[1] + list(range(5, cityVenues_merged2.shape[1]))]]

Now lets examine cluster 1


Borough  Labels for city cluster map 1st Most Common Venue      2nd Most Common Venue        3rd Most Common Venue  ...          6th Most Common Venue   7th Most Common Venue  \
0              North York                            0                   ATM       Caribbean Restaurant                Shopping Mall  ...                       Pharmacy                    Bank   
2        Downtown Toronto                            0                 Hotel                    Brewery                         Park  ...                           Café     Sporting Goods Shop   
3               Etobicoke                            0       Supplement Shop                   Pharmacy               Baseball Field  ...                         Museum            Music School   
4             Scarborough                            0  Fast Food Restaurant                Pizza Place                     Pharmacy  ...                   Skating Rink             Supermarket   
5              North York                            0   Japanese Restaurant                 Playground                  Pizza Place  ...                         Bakery                     ATM   
6               East York                            0          Skating Rink                        ATM                 Intersection  ...             Athletics & Sports                    Park   
7        Downtown Toronto                            0           Coffee Shop                Pizza Place                         Café  ...           Caribbean Restaurant          Sandwich Place   
9            East Toronto                            0           Pizza Place                        Bar                  Coffee Shop  ...                           Park     Japanese Restaurant   
10       Downtown Toronto                            0                  Café        Japanese Restaurant                        Hotel  ...             Seafood Restaurant              Restaurant   
11                   York                            0         Deli / Bodega                       Bank               Private School  ...                  Movie Theater                   Diner   
12            Scarborough                            0  Fast Food Restaurant             Discount Store               Clothing Store  ...              Indian Restaurant       Electronics Store   
14       Downtown Toronto                            0                 Hotel                Coffee Shop          Japanese Restaurant  ...                           Café              Restaurant   
15       Downtown Toronto                            0     Korean Restaurant                Coffee Shop                         Café  ...                   Dessert Shop           Grocery Store   
16             North York                            0    Italian Restaurant          Indian Restaurant           Mexican Restaurant  ...                         Bakery                    Café   
17              East York                            0     Indian Restaurant         Turkish Restaurant                  Coffee Shop  ...                   Burger Joint             Supermarket   
18           West Toronto                            0                   Bar                     Bakery                  Coffee Shop  ...                           Café            Cocktail Bar   
19            Scarborough                            0           Coffee Shop                        Gym                          Pub  ...                       Pharmacy    Fast Food Restaurant   
20  East YorkEast Toronto                            0           Pizza Place                   Pharmacy                     Bus Line  ...                           Café          Sandwich Place   
21             North York                            0                  Bank             Sandwich Place           Persian Restaurant  ...                 Breakfast Spot    Fast Food Restaurant   
24             North York                            0                  Café  Middle Eastern Resta

This cluster has 35 Neighbouhoods (red dots on the map). The most popular veunes are restaurants, hotels and cafes. 
Is reasonable to assume that looking at this table we can see what kinds of venues we can find the city centre

In [99]:
cityVenues_merged2.loc[cityVenues_merged2['Labels for city cluster map'] == 1, cityVenues_merged2.columns[[1] + list(range(5, cityVenues_merged2.shape[1]))]]

Borough  Labels for city cluster map       1st Most Common Venue      2nd Most Common Venue       3rd Most Common Venue  ...          6th Most Common Venue    7th Most Common Venue  \
1         North York                            1                    Bus Line  Middle Eastern Restaurant            Asian Restaurant  ...           Outdoor Supply Store                   Museum   
13         East York                            1                        Park             Sandwich Place         Japanese Restaurant  ...                     Nail Salon             Neighborhood   
23        North York                            1  Construction & Landscaping                     Bakery                 Flower Shop  ...  Paper / Office Supplies Store               Nail Salon   
32   Central Toronto                            1                        Park                 Playground                Home Service  ...                    Pizza Place           Pilates Studio   
33         Etobicoke                            1                        Park                Coffee Shop          Italian Restaurant  ...                   Neighborhood  New American Restaurant   
40   Central Toronto                            1                        Park                 Playground                 Candy Store  ...           Outdoor Supply Store             Music School   
41  Downtown Toronto                            1                        Park                 Playground  Construction & Landscaping  ...                   Music School               Nail Salon   

      8th Most Common Venue    9th Most Common Venue   10th Most Common Venue  
1              Music School               Nail Salon             Neighborhood  
13  New American Restaurant                Nightclub             Noodle House  
23             Neighborhood  New American Restaurant                Nightclub  
32             Music School               Nail Salon             Neighborhood  
33                Nightclub             Noodle House                   Office  
40               Nail Salon             Neighborhood  New American Restaurant  
41             Neighborhood  New American Restaurant                Nightclub  

[7 rows x 12 columns]

The second cluster has 7 neighbourhoods (blue dots)

Here we can see what kind of venues there are in recreational areas

In [101]:
cityVenues_merged2.loc[cityVenues_merged2['Labels for city cluster map'] == 3, cityVenues_merged2.columns[[1] + list(range(5, cityVenues_merged2.shape[1]))]]

Borough  Labels for city cluster map 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  ... 6th Most Common Venue  7th Most Common Venue 8th Most Common Venue  \
8              York                            3                 Field        Soccer Stadium               Dog Run  ...          Hockey Arena  Outdoors & Recreation            Nail Salon   
27  Central Toronto                            3          Skating Rink                 Trail                   ATM  ...          Music School             Nail Salon          Neighborhood   

      9th Most Common Venue   10th Most Common Venue  
8              Neighborhood  New American Restaurant  
27  New American Restaurant                Nightclub  

[2 rows x 12 columns]

This cluster has only 2 neighbouhoods (light yellow dot on the maps). This cluster indicates the venues around stadiums

In [100]:
cityVenues_merged2.loc[cityVenues_merged2['Labels for city cluster map'] == 2, cityVenues_merged2.columns[[1] + list(range(5, cityVenues_merged2.shape[1]))]]

Borough  Labels for city cluster map 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  ...          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
22  East Toronto                            2             Locksmith          Intersection             Nightclub  ...  Paper / Office Supplies Store          Music School            Nail Salon   

   9th Most Common Venue   10th Most Common Venue  
22          Neighborhood  New American Restaurant  

[1 rows x 12 columns]

This cluster has only 1 neighbouhood (cian dot on the map)